sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SpanProcessor.java

```diff
@@ -85,6 +85,32 @@ static SpanProcessor composite(Iterable<SpanProcessor> processors) {
    */
   boolean isEndRequired();
 
+  /**
+   * Called just before a {@link io.opentelemetry.api.trace.Span} is ended, if the {@link
+   * Span#isRecording()} returns true. This means that the span will still be mutable. Note that the
+   * span will only be modifiable synchronously from this callback, concurrent modifications from
+   * other threads will be prevented.
+   *
+   * <p>This method is called synchronously on the execution thread, should not throw or block the
+   * execution thread.
+   *
+   * <p>Note: This method is experimental and might be subject to future changes.
+   *
+   * @param span the {@code Span} that is just about to be ended.
+   */
+  default void onEnding(ReadWriteSpan span) {}
```

Comment:
I think we haven't had the case before where new experimental methods were added to a plugin interface.

Should we add those as default methods (like done here) or should I separate them into a new `ExperimentalSpanProcessor extends SpanProcessor` interface? If yes, do you ahve any suggestions for a package to place that interface in?

Reply:
adding the default methods to the existing interface is the right way to go. 👍🏽

Reply:
The lock was actually intentionally hold to cover this part of the spec:
> The SDK MUST guarantee that the span can no longer be modified by any other thread
before invoking `OnEnding` of the first `SpanProcessor`. From that point on, modifications
are only allowed synchronously from within the invoked `OnEnding` callbacks.

This part was added to the spec so that SpanProcessors can be sure that the state they see the span in is actually consistent and not prone to race conditions due to span modifications from the application code.

However, I now switched to a different way of implementing this, because as you said exposing the lock is probably not the best way of doing it here.

Reply:
hmm. I had missed that this is an "experimental" method. I'd like @jack-berg to chime in, but I don't think we want to be adding this at all until it's stabilized by the spec. Unless the purpose of this PR is to just demonstrate an implementation.

Reply:
Feels a bit like a chicken-egg problem: Without usage, we'll likely won't get the `OnEnding` callback to be stable soon. Without an usable implementation, we won't get usage.

That's why I was proposing to add the method to an `ExperimentalSpanProcessor extends SpanProcessor` instead:
This offers an easy path for removal/changes if the callback is removed from the spec or changed.
If it is promoted to stable, we can just move the `OnEnding` method as a default method to `SpanProcessor` and keep the `ExperimentalSpanProcessor` around for some time as empty, deprecated interface to not break existing code using it.

Reply:
I doubt adding it to a new Experimental* class will get much usage of it. For example, it wouldn't be something that would be added to the agent, which is the most common usage. What are you hoping to show with pushing it into the mainline vs. just demonstrating via a non-merged PR? 

We definitely can't add this to the primary interface until it is stable in the spec. And, I suspect creating a nearly-unused experimental class won't really demonstrate "real world" usage of the new feature. Who do you expect to use it, and for what use-case?

Reply:
> Who do you expect to use it, and for what use-case?

For contrib-extensions (e.g. the [span-stacktrace SpanProcessor](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/span-stacktrace) which was provided as a motivator for this change). Additionally there have been asks from the community (e.g. via slack) for this feature. If it were merged, they could use it at the risk of it being changed or removed of course. I thought that this was the intention behind the `Experimental` state: To have something to ship, but which is not guaranteed to be stable or even preserved in future versions.

Anyway, let's wait on what @jack-berg thinks, as he was the sponsor of this spec-change. I'm happy to follow whatever you maintainers think is the best process forward here.

Reply:
I like the idea of an extended `SpanProcessor` interface in the `io.opentelemetry.sdk.trace.internal` package. I would name it `ExtendedSpanProcessor` to mirror the `Extended*` naming pattern we've been following in the `opentelemetry-api-incubator`.

Implementations of `SpanProcessor` continue to compile, and users can choose to implement the new `ExtendedSpanProcessor` if they're willing to take a dependency on an API without backwards compatibility guarantees. If all things go well, the new `onEnding` method stabilizes, we promote it to `SpanProcessor` with a default implementation, and remove `ExtendedSpanProcessor`.

Reply:
Implemented in https://github.com/open-telemetry/opentelemetry-java/pull/6367/commits/78e715bac707958527784260282abfd9a95f2d22